### DISCUSSION TITLE
Raspbian issues

### DISCUSSION THREAD

**Discussion by alkisg** [ltsp/ltsp#261] [Q&A]
I'll track here some issues that may happen on Raspbian, but are caused by Raspbian and not by LTSP.
You may add your own issues here along with their workarounds, but please start separate issues if discussion is needed.

**Comment by alkisg** [comment]
#### Problem with the v4l2 driver

Netbooting Raspbian hangs on rpi2 and rpi3 a bit before lightdm starts. This is caused by the Video For Linux Raspberry Pi module, bcm2835_v4l2. It has been reported many times, for example https://github.com/raspberrypi/linux/issues/3067#issuecomment-576028015 and https://github.com/raspberrypi/linux/issues/2335#issuecomment-576011598.

A workaround is to add this to cmdline.txt:

```shell
modprobe.blacklist=bcm2835_v4l2
```

Edit: this has been resolved in Feb 2020.

**Comment by alkisg** [comment]
#### Icons missing

You know how I prefer maintaining images on normally booted servers/VMs/clients and not via the chroot command... this is another related issue! If one runs `chroot /srv/ltsp/raspbian apt full-upgrade`, and the **raspberrypi-ui-mods** package happens to get updated, then on the next boot all icons will be replaced by the missing icon.

I think its postinst is trying to look in some directory like /proc which isn't mounted at that time, and ends up destroying the icon cache. The solution is to netboot the administrative Raspberry in NFS-RW mode and execute the following command:

```shell
apt install --reinstall raspberrypi-ui-mods
```

**Comment by lurch** [comment]
> he raspberrypi-ui-mods package happens to get updated, then on the next boot all icons will be replaced by the missing icon.

I wonder if that might have anything to do with https://github.com/RPi-Distro/raspberrypi-ui-mods/issues/17 ?  :man_shrugging:

**Comment by alkisg** [comment]
The problem is in raspberrypi-ui-mods.postinst; it's trying to migrate settings for the active user and to detect if Xorg is running in order to display a zenity message.

It's extremely broken and it violates Debian policy; there's not much point in trying to pinpoint the exact line that fails in this specific case, as it should all be rewritten from scratch properly...

**Comment by lurch** [comment]
I'm sure @spl237 would welcome Pull Requests :slightly_smiling_face: 
But in general, I believe Raspbian favours a "user-friendly over Debian purity" approach.

**Comment by alkisg** [comment]
#### Squashfs image compression

The Raspberry Pi CPU is rather slow. I did the following benchmarks on an rpi3:

| command | seconds |
|--|--|
| cat /mnt/raspbian.img > /dev/null | 82 |
| time tar cf - /mnt/loopmounted-raspbian-img-gzip/ \| cat > /dev/null | 469 |
| time tar cf - /mnt/loopmounted-raspbian-img-lzo/ \| cat > /dev/null | 156 |

The raspbian.img (gzip) and raspbian.img.lzo were 1.5 GB.
The first command shows that transferring 1.5 GB over the network on that setup needs 82 sec.
The second command shows that transferring 1.5 GB AND decompressing 1.5 GB gzip data needs 469 sec.
The third command shows that lzo decompresses 5 times faster than gzip on an rpi3.

I think it's clear that Raspbian LTSP installations should run ` ltsp image raspbian -m '-comp lzo'`.
The equivalent ltsp.conf directive is:

```shell
[server]
MKSQUASHFS_PARAMS="-comp lzo"
```

**Comment by alkisg** [comment]
@lurch, sorry, I have no time to send pull requests there, I barely found enough time to add Raspbian support in LTSP. :)

AFAIK the proper time to do settings migration is not on postinst but at the next time the program runs, whether that is evolution, thunderbird, firefox, lxpanel etc. Wrappers can be introduced whenever necessary, even session wrappers that first check for necessary settings migration and then chain to the real session.

Otherwise it will only work for a single Xorg user case, it won't work for multiple users or SSH users etc. For example, update-manager has an /etc/xdg/autostart wrapper to notify the users that log in graphically, it doesn't try to run zenity from postinst.

But let's please start new community issues when discussion is needed, and let's keep this issue as an "index", a master page that can be read quickly by LTSP Raspbian users. Start a new issue about raspberrypi-ui-mods and I'll reply there. Thanks!

**Comment by alkisg** [comment]
By default, Raspbian file manager automounts the SD card partitions.
But if we only have bootcode.bin there, we don't want it to show up as mounted.
To disable this behavior: https://raspberrypi.stackexchange.com/questions/65739/stock-pi3-what-process-is-auto-mounting-removable-media
